gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/transactions/SnapshotTxGIIDUnit.java

```diff
@@ -0,0 +1,399 @@
+    for (int i = 0; i < numValues; i++) {
+      r.put(i, i);
+    }
+    r.getCache().getCacheTransactionManager().commit();
+  }
+
+  public void testSnapshotGII_concurrentWrites() throws Exception {
+
+    startVMs(0, 2);
+    Properties props = new Properties();
+    final Connection conn = TestUtil.getConnection(props);
+    conn.createStatement();
+
+    VM server1 = this.serverVMs.get(0);
+    VM server2 = this.serverVMs.get(1);
+    stopVMNums(-2);
+
+    server1.invoke(SnapshotTxGIIDUnit.class, "createPR", new Object[]{regionName, 2, 1});
+
+    // Put some values to initialize buckets
+    server1.invoke(new SerializableRunnable() {
+      @Override
+      public void run() {
+        putSomeValues(500);
+      }
+    });
+
+    blockGII(-2, InitialImageOperation.GIITestHookType.BeforeRequestRVV);
+
+    restartServerVMNums(new int[]{2}, 0, null, null);
+
+    server2 = this.serverVMs.get(1);
+
+    AsyncInvocation ai2 = server1.invokeAsync(new SerializableRunnable() {
+      @Override
+      public void run() {
+        putSomeValues(500);
+      }
+    });
+
+    // Create the PR region
+    server2.invoke(SnapshotTxGIIDUnit.class, "createPR", new Object[]{regionName, 2, 1});
+
+    waitForGIICallbackStarted(-2, InitialImageOperation.GIITestHookType.BeforeRequestRVV);
+
+    unblockGII(-2, InitialImageOperation.GIITestHookType.BeforeRequestRVV);
+
+    // Wait for the PR to get initialized and GII completes on the lone bucket
+    waitForRegionInit(-2);
+
+    ai2.join();
+
+    RegionVersionVector rvv1 = getRVV(-1);
+    RegionVersionVector rvv2 = getRVV(-2);
+
+    if(!rvv1.logicallySameAs(rvv2)) {
+      fail("RVVS don't match. provider=" +rvv1.fullToString() + ", recipient=" + rvv2.fullToString());
+    }
+  }
+
+  public void testSnapshotGII_killDestination() throws Exception {
+
+    startVMs(0, 2);
+    Properties props = new Properties();
+    final Connection conn = TestUtil.getConnection(props);
+    conn.createStatement();
+
+    VM server1 = this.serverVMs.get(0);
+    VM server2 = this.serverVMs.get(1);
+    stopVMNums(-2);
+
+    server1.invoke(SnapshotTxGIIDUnit.class, "createPR", new Object[]{regionName, 2, 1});
+
+    // Put some values to initialize buckets
+    server1.invoke(new SerializableRunnable() {
+      @Override
+      public void run() {
+        putSomeValues(500);
+      }
+    });
+
+    blockGII(-1, InitialImageOperation.GIITestHookType.AfterGIILock);
+
+    restartServerVMNums(new int[]{2}, 0, null, null);
+
+    server2 = this.serverVMs.get(1);
+
+
+    // Create the PR region
+    server2.invoke(SnapshotTxGIIDUnit.class, "createPR", new Object[]{regionName, 2, 1});
+
+    waitForGIICallbackStarted(-1, InitialImageOperation.GIITestHookType.AfterGIILock);
+
+    stopVMNums(-2);
+
+    unblockGII(-1, InitialImageOperation.GIITestHookType.AfterGIILock);
+
+
+
+    server1.invoke(new SerializableRunnable() {
+      @Override
+      public void run() {
+        putSomeValues(500);
+      }
+    });
```

Comment:
Sure, Will add the test.  The final set of puts are to ensure , puts does not hang.